odbc/src/api.rs

```diff
@@ -1,17 +1,83 @@
+use crate::handles::{
+    Connection, ConnectionState, Env, EnvState, MongoHandle, Statement, StatementState,
+};
 use odbc_sys::{
     BulkOperation, CDataType, Char, CompletionType, ConnectionAttribute, Desc, DriverConnectOption,
     EnvironmentAttribute, FetchOrientation, HDbc, HDesc, HEnv, HStmt, HWnd, Handle, HandleType,
     InfoType, Integer, Len, Nullability, ParamType, Pointer, RetCode, SmallInt, SqlDataType,
     SqlReturn, StatementAttribute, ULen, USmallInt, WChar,
 };
+use std::sync::RwLock;
 
 #[no_mangle]
 pub extern "C" fn SQLAllocHandle(
-    _handle_type: HandleType,
-    _input_handle: Handle,
-    _output_handle: *mut Handle,
-) -> SqlReturn {
-    unimplemented!()
+    handle_type: HandleType,
+    input_handle: Handle,
+    output_handle: *mut Handle,
+) -> SqlReturn {
+    match sql_alloc_handle(handle_type, input_handle as *mut _, output_handle) {
+        Ok(_) => SqlReturn::SUCCESS,
+        Err(_) => SqlReturn::INVALID_HANDLE,
+    }
+}
+
+fn sql_alloc_handle(
+    handle_type: HandleType,
+    input_handle: *mut MongoHandle,
+    output_handle: *mut Handle,
+) -> Result<(), ()> {
+    match handle_type {
+        HandleType::Env => {
+            let env = RwLock::new(Env::with_state(EnvState::Allocated));
+            let mh = Box::new(MongoHandle::Env(env));
+            unsafe {
+                *output_handle = Box::into_raw(mh) as *mut _;
+            }
+            Ok(())
+        }
+        HandleType::Dbc => {
+            // input handle cannot be NULL
+            if input_handle.is_null() {
+                return Err(());
+            }
+            // input handle must be an Env
+            let env = unsafe { (*input_handle).as_env()? };
+            let conn = RwLock::new(Connection::with_state(
+                input_handle,
+                ConnectionState::Allocated,
+            ));
+            let mut env_contents = (*env).write().unwrap();
+            let mh = Box::new(MongoHandle::Connection(conn));
+            let mh_ptr = Box::into_raw(mh) as *mut _;
```

Comment:
You are correct here. I think I needed that in prior code and didn't check to see if it was necessary after I changed the code (it was probably a `Handle` instead of a `MongoHandle` at some point).